Manually implement tempname on non-windows

vchuravy opened the pull request:
Fixes https://github.com/JuliaLang/julia/issues/35785

vchuravy committed:
Unset TMPDIR for tempname

musm commented on base/file.jl:
```diff
@@ -578,10 +578,14 @@ else # !windows
-    systemerror(:tempnam, p == C_NULL)
-    s = unsafe_string(p)
-    Libc.free(p)
+    # If TEMPDIR is set tempname will ignore `parent` as an argument (#38873)
```
```suggestion
    # If TMPDIR is set tempname will ignore `parent` as an argument (#38873)
```

musm commented on base/file.jl:
```diff
@@ -578,10 +578,14 @@ else # !windows
-    s = unsafe_string(p)
-    Libc.free(p)
+    # If TEMPDIR is set tempname will ignore `parent` as an argument (#38873)
+    s = withenv("TMPDIR" => nothing) do
```
shouldn't this be conditional on `parent != tempdir()` ?

vchuravy replied:
See https://github.com/JuliaLang/julia/pull/38879#issuecomment-744735667

IMO it is fine to do this unconditional. We unset `TMPDIR` here after checking it with `tempdir`, so that the libc implementation will not override `parent` with the value of `TMPDIR`

staticfloat replied:
I have a mortal fear of messing with environment variables now.  I'm pretty sure that with this change, we can crash Julia on glibc if we call `tempanme()` in parallel.

I think the only way we can avoid threading issues here is to un-set `TMPDIR` at startup and emulate it. :(

vchuravy replied:
At this point I am almost thinking it would be better to avoid the libc implementation of tempname altogether.

stevengj replied:
Should `withenv` have a mutex lock?

StefanKarpinski replied:
Avoiding `libc` is probably better. Anything that requires modifying the environment is going to be broken.

stevengj replied:
The [musl implementation](https://git.musl-libc.org/cgit/musl/tree/src/temp) seems pretty straightforward, and is MIT-licensed if we want to copy it.

StefanKarpinski commented:
Shouldn't this do the opposite? If `parent` is passed, `TMPDIR` is ignored? If someone has explicitly asked for a temporary path to be in a certain directory, then we should respect that. They might do this, for example, if they need to ensure that it is on the same file system as a final location that they need to be able to move it to without copying. With this change, setting `TMPDIR` would wreck all attempts to do that.

vchuravy commented:
> Shouldn't this do the opposite? If `parent` is passed, `TMPDIR` is ignored? If someone has explicitly asked for a temporary path to be in a certain directory, then we should respect that. They might do this, for example, if they need to ensure that it is on the same file system as a final location that they need to be able to move it to without copying. With this change, setting `TMPDIR` would wreck all attempts to do that.

I am not sure I understand. This is doing exactly what you are describing. 

If the user passes `parent` we will use that as the directory to prefix the name with, we ensure that `libc` honors the passed `parent` directory by unsetting `TMPDIR` since otherwise it would ignore `parent` and use `TMPDIR` instead. Since we use `parent=tempdir()` in the case that `parent` is not passed we still honor `TMPDIR` as an environment variable. 

We could also achieve the same by setting `TMPDIR` to `parent`, and could then pass an empty string to libc.

StefanKarpinski commented:
The comment says:

> If TMPDIR is set tempname will ignore `parent` as an argument (#38873)

That's the opposite of what it should do: if `parent` is passed, ignore `TMPDIR`.

vchuravy commented:
Currently:

```
vchuravy@odin ~> TMPDIR=~ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> tempname()
"/home/vchuravy/jl_Z9se0A"

julia> tempname("/opt")
"/home/vchuravy/jl_hI1x9x"
```

With this PR:
```
vchuravy@odin ~/s/julia (vc/tempname)> TMPDIR=~ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.3 (2020-11-09)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> @eval Base.Filesystem function tempname(parent::AbstractString=tempdir(); cleanup::Bool=true)
                  isdir(parent) || throw(ArgumentError("$(repr(parent)) is not a directory"))
                  # If TEMPDIR is set tempname will ignore `parent` as an argument (#38873)
                  s = withenv("TMPDIR" => nothing) do
                      p = ccall(:tempnam, Cstring, (Cstring, Cstring), parent, temp_prefix)
                      systemerror(:tempnam, p == C_NULL)
                      s = unsafe_string(p)
                      Libc.free(p)
                      return s
                  end
                  cleanup && temp_cleanup_later(s)
                  return s
              end
tempname (generic function with 2 methods)

julia> tempname()
"/home/vchuravy/jl_u70Ddi"

julia> tempname("/opt")
"/opt/jl_WNbezj"

```

vchuravy commented:
> The comment says:
> 
> > If TMPDIR is set tempname will ignore `parent` as an argument (#38873)
> 
> That's the opposite of what it should do: if `parent` is passed, ignore `TMPDIR`.

Ah! I see where the confusion comes from. I was documenting the behavior of `:tempnam` and not the behavior of the Julia function `tempname`.

StefanKarpinski commented:
Ah, I understand the confusion: the comment is explaining why we need to unset `TMPDIR` whereas I read it as explaining how the function behaves. Perhaps there's some clearer way of wording that? Something like this:

> Unless we unset `TMPDIR` here, it will cause `tempnam` to give a directory in that location, which would contradict the `parent` argument passed in here.

I worry about the thread safety of messing with `TMPDIR` like this. What if, for example, two different threads do this at the same time?

vchuravy commented:
> I worry about the thread safety of messing with TMPDIR like this. What if, for example, two different threads do this at the same time?

Yeah... `setenv` is not thread-safe at all.

StefanKarpinski commented:
We could at least put a lock around this so that multiple calls to this function from different threads don't end up concurrently modifying the same environment variable.

DilumAluthge commented:
Instead of `setenv`, would `withenv` work? Would that be thread-safe?

vchuravy commented:
> Instead of `setenv`, would `withenv` work? Would that be thread-safe?

I am using `withenv` here, but that uses `setenv` internally.

DilumAluthge commented:
> > Instead of `setenv`, would `withenv` work? Would that be thread-safe?
> 
> I am using `withenv` here, but that uses `setenv` internally.

Ah fair enough

staticfloat commented:
The issue is not just Julia code; if any C library (such as, for instance, `tempname()`) tries to read an environment variable while another C library (or some Julia task) modifies the environment block, we can segfault.  We don't run into this very often because we tend to set environment variables at the beginning of program execution and leave them from then on, but the more we use `withenv()` the more likely we are to create this problem.  To be explicit, a Julia mutex won't do anything but ensure that Julia can't interfere with other Julia operations.  That's not even as helpful as you might think however, because since `tempnam()` uses `getenv()` to read the contents of `$TMPDIR`, if you have two Julia tasks that each wait for a mutex to call `setenv()` or `getenv()`, you can still have Julia acquire the (julia) mutex before calling `setenv()` and have that occur at a bad point in time for some other Julia task which is calling `tempnam()`.

Possible solutions:
* We could surround all `ccall()`'s with acquiring a mutex and not release it until the `ccall()` returns.  Ha ha, right.  This at least avoids crashes being Julia's fault; C code could still start a thread that calls `setenv()` within it at some point and causes a segfault.

* We can hijack `getenv()` and `setenv()` somehow and force these functions to become threadsafe.  A custom-build libc, symbol injection, etc...  This is the only truly foolproof method that I know of, and will impose a performance penalty on environment actions, but IMO that's totally acceptable.

* We can avoid modification of environment variables at unsafe points (where "unsafe" means after any thread has been started by anything).  Realistically, this means you should set environment variables in things like `__init__()` and leave it at that.  I will note that I am guilty as anyone in that JLL wrappers set environment variables when launching external processes, but I've been working to eliminate that by passing an environment block to `open(::Cmd)` rather than using `withenv`, and in so doing avoiding `setenv()` on the parent process altogether.

Keno reviewed (dismissed):
Per discussion, should avoid using libc, since messing with the environment is inherently problematic.

vchuravy committed:
implement tempname ourselve

vchuravy commented on base/file.jl:
```diff
@@ -578,21 +578,40 @@ end
 
+# Based of musl __randname
+function _rand_string()
+    r = Base.rand(UInt) # FIXME: https://git.musl-libc.org/cgit/musl/tree/src/temp/__randname.c
```
Any good ideas on how to get a random number here? Musl use `time_ns`, but then mixes in two pointers as sources of randomness.

musm replied:
Is there a particular reason we are avoiding `Base.rand` ?

stevengj replied:
Even if we use `Base.rand`, we might want to initialize our own seed since we don't want the caller to be able to override this, no?

vchuravy replied:
>  since we don't want the caller to be able to override this, no?

I don't think we care about the state of the rng. IIUC we just want a random seed that is not likely to be used by another process. This is why musl mixes in the two pointers to get some randomness in the higher bits, but we don't need to high quality randomness. 

> Is there a particular reason we are avoiding `Base.rand` ?

`Base.rand` is implemented through the stdlib `Random` and thus we would add a hard-dependency edge. Otherwise I would just use `Random.randstring`.

vtjnash replied:
I believe `Libc.rand()` should provide suitable quality for this. It appears we can also use the same code everywhere now, rather than using separate, identical implementation for windows.

musm commented on base/file.jl:
```diff
@@ -578,15 +578,40 @@ end
+end
+
 # Obtain a temporary filename.
 function tempname(parent::AbstractString=tempdir(); cleanup::Bool=true)
```
We should use this implementation on Windows as well

musm commented:
I find it a bit odd we are trying to implement `tempnam`  since  the first sentence of the description section in http://man7.org/linux/man-pages/man3/tempnam.3.html says "Never use this function." ... use `mkstemp` instead.

vchuravy commented:
> I find it a bit odd we are trying to implement `tempnam` since the first sentence of the description section in http://man7.org/linux/man-pages/man3/tempnam.3.html says "Never use this function." ... use `mkstemp` instead.

:shrug: I read that as well and found it entertaining, I assume they want to discourage usage since there is the risk of running into collisions and `mkstemp` will lock the file immediatly.

StefanKarpinski commented:
OMG, can we finish this?

vchuravy commented:
> OMG, can we finish this?

xD If anyone want to pick this up, please go ahead.

staticfloat committed:
Unify windows/non-windows `tempname()` implementations

staticfloat commented:
Rebased and unified a bit, as requested by reviewers.

stevengj commented on base/file.jl:
```diff
@@ -554,8 +554,41 @@ end
+# Use `Libc.rand()` to generate random strings
+function _rand_filename(len = 10)
+    slug = Base.StringVector(len)
+    chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
```suggestion
    chars = b"0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
You really want an array of bytes here, I think.

If `chars` is a string, then `chars[i]` converts the byte to a `Char` and then it gets converted back for assignment to `slug[i]`, and `length(chars)` is slow.

staticfloat replied:
Thanks @stevengj!

staticfloat replied:
This is a pretty decent speedup.  The old version:

```
julia> @btime _rand_filename(10)
  320.220 ns (2 allocations: 96 bytes)
```

The new version:
```
julia> @btime _rand_filename(10)
  66.308 ns (2 allocations: 96 bytes)
```

staticfloat committed:
Use array of bytes, rather than a string

Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

staticfloat committed:
Rename `_win_tempname()` -> `_win_mkstemp()`

This was a confusingly-named function, since `tempname()` doesn't create
a file (and is thus prone to race conditions) whereas `mkstemp()`
creates a file, reserving that filename as used immediately.

musm commented:
Any reason not to use this also on Windows?

staticfloat commented:
> Any reason not to use this also on Windows?

Despite the PR title, this does unify much of the windows/posix divide.

staticfloat commented:
This is green, and I believe all reviewer concerns have been addressed.  I'm going to add the `merge me` label.

vchuravy commented:
Thanks Elliot for picking this up!

DilumAluthge commented:
Thank you Valentin and Elliot!

Keno commented:
CI failures might be related:
https://build.julialang.org/#/builders/44/builds/856
https://build.julialang.org/#/builders/72/builds/866

Keno commented:
I will note that the file name is the same in both of these failures, so I'm assuming the RNG doesn't get seeded during precompile generation and since we're using tempname here, we probably end up clashing somewhere.

Keno commented:
We may want to revert this until #43597 is merged. This is likely quite destabilizing to CI.
